ci failures with java.util.ConcurrentModificationException

Caused by: java.lang.RuntimeException: java.util.ConcurrentModificationException
	at com.pingcap.tikv.meta.TiDAGRequest.copy(TiDAGRequest.java:1033)
	at org.apache.spark.sql.execution.RegionTaskExec$$anonfun$internalRowToUnsafeRowWithIndex$3$$anonfun$apply$3$$anon$1.call(CoprocessorRDD.scala:264)
	at org.apache.spark.sql.execution.RegionTaskExec$$anonfun$internalRowToUnsafeRowWithIndex$3$$anonfun$apply$3$$anon$1.call(CoprocessorRDD.scala:262)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	... 3 more
Caused by: java.util.ConcurrentModificationException
	at java.util.ArrayList.writeObject(ArrayList.java:770)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1140)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1496)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at com.pingcap.tikv.meta.TiDAGRequest.copy(TiDAGRequest.java:1028)
	... 8 more

--- a/core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala
+++ b/core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala
@@ -180,6 +180,7 @@ case class RegionTaskExec(child: SparkPlan,
   private val downgradeThreshold =
     sqlConf.getConfString(TiConfigConst.REGION_INDEX_SCAN_DOWNGRADE_THRESHOLD, "10000").toInt
   private lazy val project = UnsafeProjection.create(schema)
+  private val logger = Logger.getLogger(getClass.getName)
 
   type TiRow = com.pingcap.tikv.row.Row
 
@@ -215,7 +216,6 @@ case class RegionTaskExec(child: SparkPlan,
     downgradeDagRequest: TiDAGRequest
   ): (Int, Iterator[InternalRow]) => Iterator[UnsafeRow] = { (index, iter) =>
     // For each partition, we do some initialization work
-    val logger = Logger.getLogger(getClass.getName)
     logger.debug(s"In partition No.$index")
     val session = TiSessionCache.getSession(tiConf)
     session.injectCallBackFunc(callBackFunc)
@@ -261,7 +261,7 @@ case class RegionTaskExec(child: SparkPlan,
         taskCount += 1
         val task = new Callable[util.Iterator[TiRow]] {
           override def call(): util.Iterator[TiRow] =
-            CoprocessIterator.getRowIterator(dagRequest.copy(), tasks, session)
+            CoprocessIterator.getRowIterator(dagRequest, tasks, session)
         }
         completionService.submit(task)
       }

--- a/core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala
+++ b/core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala
@@ -180,7 +180,6 @@ case class RegionTaskExec(child: SparkPlan,
   private val downgradeThreshold =
     sqlConf.getConfString(TiConfigConst.REGION_INDEX_SCAN_DOWNGRADE_THRESHOLD, "10000").toInt
   private lazy val project = UnsafeProjection.create(schema)
-  private val logger = Logger.getLogger(getClass.getName)
 
   type TiRow = com.pingcap.tikv.row.Row
 
@@ -216,6 +215,7 @@ case class RegionTaskExec(child: SparkPlan,
     downgradeDagRequest: TiDAGRequest
   ): (Int, Iterator[InternalRow]) => Iterator[UnsafeRow] = { (index, iter) =>
     // For each partition, we do some initialization work
+    val logger = Logger.getLogger(getClass.getName)
     logger.debug(s"In partition No.$index")
     val session = TiSessionCache.getSession(tiConf)
     session.injectCallBackFunc(callBackFunc)

--- a/core/src/test/scala/org/apache/spark/sql/test/SharedSQLContext.scala
+++ b/core/src/test/scala/org/apache/spark/sql/test/SharedSQLContext.scala
@@ -19,11 +19,9 @@ package org.apache.spark.sql.test
 
 import java.io.File
 import java.sql.{Connection, DriverManager, Statement}
-import java.util
 import java.util.{Locale, Properties, TimeZone}
 
 import com.pingcap.tispark.TiConfigConst.PD_ADDRESSES
-import com.pingcap.tispark.TiDBOptions
 import com.pingcap.tispark.statistics.StatisticsManager
 import org.apache.spark.internal.Logging
 import org.apache.spark.sql._

--- a/tikv-client/src/main/java/com/pingcap/tikv/operation/SchemaInfer.java
+++ b/tikv-client/src/main/java/com/pingcap/tikv/operation/SchemaInfer.java
@@ -43,7 +43,7 @@ public class SchemaInfer {
   }
 
   public static SchemaInfer create(TiDAGRequest dagRequest, boolean readHandle) {
-    return new SchemaInfer(dagRequest, readHandle);
+    return new SchemaInfer(dagRequest.copy(), readHandle);
   }
 
   private SchemaInfer(TiDAGRequest dagRequest, boolean readHandle) {

--- a/tikv-client/src/main/java/com/pingcap/tikv/operation/iterator/CoprocessIterator.java
+++ b/tikv-client/src/main/java/com/pingcap/tikv/operation/iterator/CoprocessIterator.java
@@ -71,12 +71,13 @@ public abstract class CoprocessIterator<T> implements Iterator<T> {
    */
   public static CoprocessIterator<Row> getRowIterator(
       TiDAGRequest req, List<RegionTask> regionTasks, TiSession session) {
+    TiDAGRequest dagRequest = req.copy();
     return new DAGIterator<Row>(
-        req.buildTableScan(),
+        dagRequest.buildTableScan(),
         regionTasks,
         session,
-        SchemaInfer.create(req),
-        req.getPushDownType()) {
+        SchemaInfer.create(dagRequest),
+        dagRequest.getPushDownType()) {
       @Override
       public Row next() {
         if (hasNext()) {

